kcqrs: implement Snapshots

VasilMitovClouway opened the pull request:
Work in progress.
Fixes:  #11

VasilMitovClouway committed:
kcqrs: implement Snapshots

mgenov commented on kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/JsonSerializer.kt:
```diff
@@ -0,0 +1,16 @@
+/**
+ * @author Vasil Mitov <vasil.mitov@clouway.com>
+ */
+interface JsonSerializer {
```
MessageFormat is already added for these purposes.

VasilMitovClouway replied:
Will be removed.

mgenov commented on kcqrs-appengine/src/main/kotlin/com/clouway/kcqrs/adapter/appengine/AppEngineEventStore.kt:
```diff
@@ -84,11 +110,23 @@ class AppEngineEventStore(private val kind: String = "Event", private val messag
+                dataStore.put(transaction, listOf(snapshotEntity, aggregateEntity))
+                transaction.commit()
+                SaveEventsResponse.Success(aggregateId, currentVersion)
+            } catch (ex: ApiProxy.RequestTooLargeException) {
```
You can check the size of the entity by converting it to Protocol Buffer. This is going to skip the round trip to the datastore. 
```kotlin
val ep = EntityTranslator.convertToPb(entity)
ep.propertySize()
```

Currently entity size is limited to: 1,048,572 bytes

VasilMitovClouway replied:
Agreed will be implemented.

VasilMitovClouway replied:
Sadly i don't think this is a proper way to check size as the propertySize returns int and it is 1. This is actually some list size in the context of 
```Kotlin
OnestoreEntity.Property
```
which is a list and apparently this returns the size of that list.

mgenov replied:
The following snippet returns the proper size of the encoded protocol buffer:

```kotlin
  entity = Entity("test")
    entity.setProperty("test1", "test")
    entity.setProperty("test2", "teaskdjakdjajsdst2")
    entity.setProperty("test3", "test3")
val pb = EntityTranslator.convertToPb(entity)
    println(pb.serializedSize)
```

The size is changed as I change the content or the number of properties.

VasilMitovClouway replied:
I was trying with .propertySize(). Will test with serializedSize.

VasilMitovClouway replied:
Implemented.

mgenov commented on kcqrs-appengine/src/main/kotlin/com/clouway/kcqrs/adapter/appengine/AppEngineEventStore.kt:
```diff
@@ -84,11 +110,23 @@ class AppEngineEventStore(private val kind: String = "Event", private val messag
+            } catch (ex: ApiProxy.RequestTooLargeException) {
+                var snapshot: Snapshot? = null
+                //if a build snapshot does not exist it would not have the field data filled in.
+                if (snapshotEntity.getProperty("data") as Blob? != null) {
```
This code looks really strange as logic to me. 

1. HttpEventStore Client -> RPC -> AppEngineEventStore -> LimitReachedException
2. HttpEventStore Client (makes snapshot of the current aggregate) -> RPC -> AppEngineEventStore  -> creates snapshot add stores new events in the new datastore

This is how I understand the flow.

VasilMitovClouway replied:
The workflow is as follows The aggregate repository tries to save an entity (e.g add the new events) 
the EventStore returns that it has reached it's limit with an SaveEventsResponse.CurrentEventsLimitReached
```Kotlin 
data class CurrentEventsLimitReached(val currentEvents: List<EventPayload>, val currentSnapshot: Snapshot? = null) : SaveEventsResponse()
```
As i see the object is build from the events in the SimpleAggregateRepository, that's why i've chosen to return the current events and a snapshot if available back to the SimpleAggregateRepository. 
The fact that the current events limit has been reached prompts the SimpleAggregateRepository to retry it's request with the same SaveOptions but including a new setting 

```Kotlin
data class CreateSnapshot(val required: Boolean = false, val snapshotData: String = "")
```
with required set to true and the newly build snapshot.

Could be optimized with the snapshotData being a byte array altogether. 

The new snapshot is being build from the old snapshot ( if available) and the current events that have reached their limit but without the new ones. This way the new events will not be digested in the new snapshot and will be present in the new entity ( or whatever persistence we have chosen)

Then sending the new snapshot with the events that caused the event limit reach and saving them in a new entity. 

This way the SimpleAggregateRepository retries the save of the events by itself and manages the snapshot creation.

mgenov commented on kcqrs-core/build.gradle:
```diff
@@ -26,6 +26,9 @@ configurations {
   compile "org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlin_version"
+  compile group: 'com.google.code.gson', name: 'gson', version: '2.7'
+
+  compile group: 'com.esotericsoftware', name: 'kryo', version: '4.0.0'
```
Not able to see where kryo is used. I'm not sure that the core module should depend on this ones. We already have gson module, so if kryo is appropriate for this it would be better to be added as separate module.

VasilMitovClouway replied:
Was only testing around. 
I ran into a problem with Gson and just today have i found what it was. All the Aggregates extend the AggregateRootBase. I constantly couldn't deserialize the snapshot as it would say that it could not be deserialized to AggregateRootBase. 

Today i've found that the problem actually comes from the interface Event. 
Simple solution would be to use Gson's annotation 
```Kotlin
  @Expose
```
Setting all of aggregate root base's fields with it except the changes list. I think changes will not be needed in the building of the Aggregate as it will use the current version of the aggregate and just apply the new events.

VasilMitovClouway replied:
And adding new methods for serialization and deserialization ( so that the old ones would not be affected) using a Gson with the setting to use annotations while serializing and deserializing.

mgenov replied:
I'm a bit afraid of such usage (Direct Serialisation of AggregateRoot) as everyone could use it's own modifiers and access to the object which may cause and issues. Enforcing of such thing will require adding of tests for these serializations which are targeting these objects.

VasilMitovClouway commented:
I ran into a problem that made me loose a lot of time. When snapshots were created the constructor that was used was applying the created event which was messing up the logic. I have not implemented a default snapshotmapper leaving the method to be forced to be implemented by the aggregates.

mgenov commented:
Yes, the state cannot be set through the same constructor. Either fields need to be set using specific factory method/constructor or through some reflection magic by using the default constructor.

VasilMitovClouway commented:
I've fixed it with marking the changes as commited after the snapshot is created. After that the snapshot has the history applied over it

mgenov commented:
I'm not sure that this is going to work as expected:


Here is a sample constructor:
```kotlin
    constructor(id: String, customerName: String) : this(customerName) {
        applyChange(OrderCreatedEvent(id, customerName))
    }
```

This is the body of applyChange from master: 

```kotlin
private fun applyChange(event: Event, isNew: Boolean) {

    var method: Method? = null

    try {
      method = this::class.java.getDeclaredMethod("apply", event::class.java)
    } catch (e: NoSuchMethodException) {
      //do nothing. This just means that the method signature wasn't found and
      //the aggregate doesn't need to apply any state changes since it wasn't
      //implemented.
    }

    if (method != null) {
      method.isAccessible = true
      try {
        method.invoke(this, event)
      } catch (e: IllegalAccessException) {
        throw IllegalStateException(e)
      } catch (e: IllegalArgumentException) {
        throw IllegalStateException(e)
      } catch (e: InvocationTargetException) {
        throw IllegalStateException(e)
      }
    }

    if (isNew) {
      changes.add(event)
    }
```

e.g apply functions are invoked immediately.

VasilMitovClouway commented:
The actual problem wasn't that the apply is called. The problem is that every time the aggregate was build, it had the initial created event also added to it. When it had the "currentEvents" ( the events so far) applied over it with loadFromHistory that initial event was left with the aggregate and send to the InMemoryEventStore. It was a strange behavior for me.

mgenov commented:
As I can see from the Order in the KCQRS lib:

```kotlin
internal class Order private constructor(var customerName: String) : AggregateRootBase() {
    @Suppress("UNUSED")
    constructor() : this("")

    constructor(id: String, customerName: String) : this(customerName) {
        applyChange(OrderCreatedEvent(id, customerName))
    }

    fun updateCustomer(newName: String) {
        applyChange(CustomerUpdatedEvent(newName))
    }

    ....
}
```

There is a default constructor which could be used for the snapshots. 

```kotlin
val aggregate = clazz.newInstance()
Snapshot.applyTo(aggregate) // set fields using reflection 
```

VasilMitovClouway commented:
I do not like this suggestion. The snapshot does not know about the snapshot mapper as the mapper is part of the AggregateRoot and the Snapshot is part of the EventStore. If i am to implement this, it means that i would have to add a dependency between the two which i do not think is needed (about the domain method applyTo) Maybe i could do this at the place im doing the 
```Kotlin
aggregate.markChangesAsCommitted()
```
And do something like 
```Kotlin 
var aggregate = clazz.newInstance()
aggregate = aggregate.fromSnapshot(snapshot)
```
in AggregateRootBase 
```
fun fromSnapshot(snapshotData:String) {
val instance = getSnapshotMapper().fromSnapshot(snapshotData)
//do the reflection logic here
}

VasilMitovClouway commented:
I've managed to add a default snapshot mapper which could be used for now. When we start adding new properties to aggregates we can override the getSnapshotMapper method and start doing the compensation logic there. Also we need to ovverride the getSnapshotMapper if we have specific fields like LocalDateTime etc.

mgenov commented:
Take a look again at my suggestion by skipping the last line.

VasilMitovClouway commented:
I've managed to apply a copy of the fields over reflection. I've added tests for the default mapper as well. I will just add an object inside an object inside my TestClass to test if that will work as intended.

VasilMitovClouway commented:
Added some more complexity to the TestObject to see how it will handle. I think PR is ready for a final review.

VasilMitovClouway commented:
Friendly reminder.

StanimirIliev commented on kcqrs-appengine/src/main/kotlin/com/clouway/kcqrs/adapter/appengine/AppEngineEventStore.kt:
```diff
@@ -40,13 +57,22 @@ class AppEngineEventStore(private val kind: String = "Event", private val messag
 
-            val aggregateIndex = snapshot.getProperty("aggregateIndex") as Long? ?: 0
+            var aggregateIndex = snapshotEntity.getProperty("aggregateIndex") as Long? ?: 0
+            var version = snapshotEntity.getProperty("version") as Long? ?: 0
```
You can made version immutable

VasilMitovClouway commented:
Pr has been tested and implemented with an example entity MyEntity. I have added some missing things in the HttpEventStore. Now when snapshot is required the client apis will need to return status code 422 (Unprocessible Entity) with a body containing the current events and the snapshot data if present.

georgievJon reviewed (changes requested):
Make test with aggregate which contains one changeable field and one field set only in first event. After several snapshots the aggregate should return correct values of fields.

VasilMitovClouway commented:
@georgievJon Scenario has been covered with the testing of the snapshot mapper.

qnislavnachev commented:
LGTM

mgenov commented on kcqrs-core/src/main/kotlin/com/clouway/kcqrs/core/AggregateRootBase.kt:
```diff
@@ -10,80 +11,126 @@ import java.util.*
+                return gson.toJson(data)
+            }
+
+            override fun fromSnapshot(snapshot: String): AggregateRoot {
```
Version of the snapshot could be passed as second parameter, so the clients will be able to distinguish between versions: 

Snapshot(payload, 1) 
Snapshot(payload, 2) 

```kotlin
if (version == 1)  {
   // read version 1 of the snapshot 
} else if (version == 2) {
  // read version 2 of the snapshot
}
```

Now by passing the payload directly, the user has to try to deserialize multiple times.

VasilMitovClouway replied:
Could be done, but how will this improve the deserialization. As i see it, the snapshot mapper will be the one responsible to handle the missing fields from previous snapshots.

VasilMitovClouway replied:
If a change to the aggregate was done, the user making the change will have to implement a snapshot mapper, different from the default, which will handle the missing fields. When a new change is made a change to the snapshot mapper will have to be done. This is a workflow that we will have to start using.

mgenov replied:
It will make life easier for the snapshot mapper when there are more then 2 versions of the snapshot. 

1. Snapshot (1)
2. Snapshot (1)
3. Snapshot (2)
4. Snapshot (2)
5. Snapshot (3)
6. Snapshot (3)

e.g if changes between version 1 and 3 are major then the mapper will be able to check before the serialization what kind of object to provide for de-serialization. 

```kotlin
if (version == 1) {
  return gson.fromJson(OrderSnapshotV1::class.java).adapt()
} else if (version == 2) {
  return gson.fromJson(OrderSnapshotV2::class.java).adapt()
} else {
  return gson.fromJson(OrderSnapshotCurrent::class.java).adapt()
}
```

e.g the client code will be able to use different types instead of nullable and optional fields.

VasilMitovClouway replied:
Added the snapshot version to both the AggregateRoot and the SnapshotMapper.

mgenov replied:
The AggregateRoot doesn't have to know about versions.

mgenov commented on kcqrs-appengine/src/main/kotlin/com/clouway/kcqrs/adapter/appengine/AppEngineEventStore.kt:
```diff
@@ -40,13 +56,22 @@ class AppEngineEventStore(private val kind: String = "Event", private val messag
+            if (saveOptions.createSnapshot.required && saveOptions.createSnapshot.snapshotData != null) {
+                aggregateIndex += 1
+                val snapshotData = saveOptions.createSnapshot.snapshotData
+                snapshotEntity.setUnindexedProperty("version", version + 1)
```
The snapshot version should not be incremented unless snapshot is changed. The only object who knows about these changes is the SnapshotMapper interface.

VasilMitovClouway replied:
We are actually sure that a new snapshot was created from the check: 
```Kotlin
if (saveOptions.createSnapshot.required && saveOptions.createSnapshot.snapshotData != null)
```
I could move the version incrementation to the snapshot mapper, but it will have to receive the current version of the snapshot as well.

mgenov replied:
I'm not able to see a reason for this as the SnapshotMapper is responsible for the creation of the snapshot and he knows what the current version is.

mgenov replied:
Check my comment below.

mgenov commented on kcqrs-core/src/test/kotlin/com/clouway/kcqrs/core/SimpleAggregateRepositoryTest.kt:
```diff
@@ -219,4 +291,103 @@ class SimpleAggregateRepositoryTest {
+class TestClassSnapshotMapper : SnapshotMapper<TestClass> {
+    private val gson = Gson()
+
+    override fun toSnapshot(data: TestClass): String {
```
My proposal was this method signature to be changed to
```kotlin
toSnapshot(data: TestClass): Snapshot
```

where the Snapshot object will only have two properties: payload and version. 


```kotlin
  // return version 1 of the snapshot
  return Snapshot(gson.toJson(mapped), 1)
```

Where this version is going to be persisted in the datastore and latter to be used for loading.

mgenov replied:
.

VasilMitovClouway replied:
I could actually reuse the existing Snapshot. What do you think.

mgenov replied:
I'm not sure that I understand what you mean.

VasilMitovClouway replied:
We have an existing class Snapshot that is accessible from here : 
```Kotlin
data class Snapshot(val version: Long, val data: Binary)
```

mgenov replied:
Yes, it could be used.

VasilMitovClouway replied:
Implemented and removed the version incrementation from the EventStore. Now it will persist the one provided by the aggregate repository.

mgenov reviewed (approved):
LGTM !

mgenov commented:
Snapshot version is released.

VasilMitovClouway commented:
Is it still under v0.0.4-SNAPSHOT

mgenov commented:
Yes, I haven't dumped new version.

On Tue, Aug 21, 2018, 09:39 Vasil Mitov <notifications@github.com> wrote:

> Is it still under v0.0.4-SNAPSHOT
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/clouway/kcqrs/pull/33#issuecomment-414566222>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAMxR4RirSivdU0JLfudQC9igzPV2dWbks5uS6sdgaJpZM4U7onT>
> .
>
